### PR TITLE
Demonstrate fix of collection typing

### LINKED ISSUE #25: Generic typing of BaseRequest, BaseStreamRequest
From what I can see BaseRequest and BaseStreamRequest should have a generic type corresponding to the responseClass. Applying this change is pretty straightforward but changes need to be made to the generated classes also. What do you suggest?
[AB#6044](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6044)

### LINKED ISSUE #31: Improve IRequestBuilder
As far as I can see, every non-abstract implementation of IRequestBuilder has these methods:
T buildRequest();
T buildRequest(final java.util.List<? extends Option> requestOptions);
where T is the return type in the implementation.
A simplification is then to move those methods to IRequestBuilder and add a generic type to IRequestBuilder corresponding to the return type (after all if there is one method that I expect to see on a RequestBuilder it is buildRequest!).
For example, this interface:
public interface IBaseAttachmentRequestBuilder extends IRequestBuilder {
    IAttachmentRequest buildRequest();
    IAttachmentRequest buildRequest(final java.util.List<? extends Option> requestOptions);
...
would become:
public interface IBaseAttachmentRequestBuilder extends IRequestBuilder<IAttachmentRequest> {
...
There are some flow on effects to abstract classes like BaseRequestBuilder but I think the generic type can be added to those classes also without issue.
[AB#6042](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6042)

### LINKED ISSUE #120: BaseRequest typing
Similar to discussion in #116, `BaseRequest` is not modelled optimally either. The obvious indicator is the very free typing of the send methods:

```java
protected <T1, T2> T1 send(final HttpMethod method,
                               final T2 serializedObject) throws ClientException;

protected <T1, T2> void send(final HttpMethod method,
                                 final ICallback<T1> callback,
                                 final T2 serializedObject) 
```
T1 and T2 should not be there, `BaseRequest` should 

* carry one generic type T (e.g. `BaseRequest<User>`) and the types represented by `responseClass`, T1, T2 should be based on T as per demo in #119.

@darrelmiller Typing-wise the codebase is a mess but I don't think this is that big a job to fix. You sort this out for one use case and code generation propagates it everywhere.

### PR DESCRIPTION
**Pull request by davidmoten**
related #116
fixes #25 
fixes #31
fixes #120 

This is what I've done:
* changed the signature of `BaseCollectionRequest` to have only one generic parameter T
* changed the constructor to expect a `responseClass` that extends `ICollectionResponse<T>`
* added `ICollectionResponse`
* `send` now returns an `ICollectionResponse<T>`
* and more that you can see

I've just addressed the T1 class in the constructor. Handling the T2 involves a little more but is similar. 

Perhaps this will help you understand the changes proposed.

### CODE REVIEW

**Comment by baywet** [comment]
Hey @davidmoten thanks for the PR and sorry for coming back to you on this one so late.
Is it still something you're looking at merging? if yes can you resolve the conflicts first please?
Thanks!

**Comment by baywet** [comment]
Here is a bit of context for the review.
Dave had put together 2 PRs that were demonstrating generics improvements but were not finished.
I merged the other one into this one, updated from feature/v3 (which itself is mostly up to date from dev) and started of that.
This improves significantly the use of generics and reduces the amount of generated code.

I couldn't go "all the way" removing things like collection pages, collection response etc... as the java generic type erasure "looses some information" if you don't have a type inheriting from the generic type.
Any commit prefixed with "codegen" is a code generation, and you can probably sample the review for those, other ones are manual changes. There's a lot of trial and error in the history, sorry about the mess.

The corresponding typewriter PR is here https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/313

The changes of this generation have been run against raptor here.
https://microsoftgraph.visualstudio.com/Graph%20Developer%20Experiences/_build/results?buildId=31836&view=results

It's complaining about the todo stuff which is normal as I haven't updated the feature/v3 branch from latest metadata in a while. Besides that raptor allowed me to catch one generation + generic issue as you can see in the latest commits.

**Review by MIchaelMainer — COMMENTED** [review]
We had a review meeting online to discuss the changes. I'll follow up after I've done some testing of these changes

File: src/main/java/com/microsoft/graph/http/BaseActionCollectionRequestBuilder.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import com.microsoft.graph.core.IBaseClient;
+import com.microsoft.graph.http.BaseRequestBuilder;
+import com.microsoft.graph.options.Option;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The base method request builder class used for POST actions
```
Comment (MIchaelMainer) [inline src/main/java/com/microsoft/graph/http/BaseActionCollectionRequestBuilder.java:16]: `The base method request builder class used for OData Action POST requests that return a collection.` 

The naming led me to believe that this is bound to a collection, and not referring to the return type.

Comment (baywet) [reply src/main/java/com/microsoft/graph/http/BaseActionCollectionRequestBuilder.java:19]: is there an action to take on that?

**Comment by baywet** [comment]
@zengin @MIchaelMainer FYI we do have the type summary working now, it contains a lot of changes but as far as I can tell this PR is not removing anything. Please have a look whenever you have some time.

**Review by zengin — APPROVED** [review]

**Review by MIchaelMainer — APPROVED** [review]

File: src/main/java/com/microsoft/graph/http/BaseActionCollectionRequestBuilder.java (hunk 1 above)
Comment (MIchaelMainer) [reply src/main/java/com/microsoft/graph/http/BaseActionCollectionRequestBuilder.java:19]: Yeah, the request is to change the comment to proposed comment.

Comment (baywet) [reply src/main/java/com/microsoft/graph/http/BaseActionCollectionRequestBuilder.java:19]:
```suggestion
 * The base method request builder class used for OData Action POST requests that return a collection.
```

Comment (baywet) [reply src/main/java/com/microsoft/graph/http/BaseActionCollectionRequestBuilder.java:19]: like this?

**Pull request merged**